File: 44. Wildcard Matching/wildcard-matching.go
// 44. Wildcard Matching (3/14/2022)
// Runtime: 8 ms (80.67%) Memory: 2.87 MB (85.71%) 

func isMatch(target string, pattern string) bool {
	target_idx, pattern_idx := 0, 0
	target_idx_retry, pattern_idx_retry := 0, 0

	for pattern_idx < len(pattern) || target_idx < len(target) {
		if pattern_idx < len(pattern) {
			c := pattern[pattern_idx]
			switch c {
			case '?':
				if target_idx < len(target) {
					pattern_idx++
					target_idx++
					continue
				}
			case '*':
				pattern_idx_retry = pattern_idx
				target_idx_retry = target_idx + 1
				pattern_idx++
				continue
			default:
				if target_idx < len(target) && target[target_idx] == c {
					pattern_idx++
					target_idx++
					continue
				}
			}
		}
        
		if 0 < target_idx_retry && target_idx_retry <= len(target) {
			pattern_idx = pattern_idx_retry
			target_idx = target_idx_retry
			continue
		}
		
		return false
	}

	return true
}

File: 44. Wildcard Matching/wildcard-matching.c
// 44. Wildcard Matching (3/14/2022)
// Runtime: 8 ms (67.92%) Memory: 5.79 MB (77.01%) 

bool isMatch(char* str, char* pat) {
    int str_len = strlen(str), pat_len = strlen(pat);
    int str_idx = 0, pat_idx = 0;
    int str_idx_retry = 0, pat_idx_retry = 0;

    while (str_idx < str_len || pat_idx < pat_len) {
        // printf("---\nstr: %s\npat: %s\n", str + str_idx, pat + pat_idx);
        if (pat_idx < pat_len) {
            if (pat[pat_idx] == '*') {
                pat_idx_retry = pat_idx;
                str_idx_retry = str_idx + 1;
                pat_idx++;
                continue;
            }
            else if (pat[pat_idx] == '?' && str_idx < str_len) {
                str_idx++, pat_idx++;
                continue;
            }
            else if (str[str_idx] == pat[pat_idx]) {
                str_idx++, pat_idx++;
                continue;
            }
        }

        if (str_idx_retry > 0 && str_idx_retry <= str_len) {
            str_idx = str_idx_retry;
            pat_idx = pat_idx_retry;
            continue;
        }

        return false;
    }

    return true;
}
